### PR TITLE
Modal

### LINKED ISSUE #9: Add component : Modal

### PR DESCRIPTION
**Pull request by bleucitron**
Added `Modal` component with tests.

`Sidebar.Title`, `Sidebar.Header`, `Sidebar.Footer` were used to add content for Modal, but this should be fixed by:
* either add `Title`, `Header` and `Footer` for `Modal`
* or make them separate components to be used in different contexts _(recommended)_

Fixes #9

### CODE REVIEW

**Comment by GuillaumeAmat** [comment]
Hi @iOiurson, thank you for your PR!

I really like your design approach, I did not think about making modals as fullscreen-ish container it's nice.

File: src/components/Modal/Modal.jsx [hunk 1]
```diff
@@ -0,0 +1,174 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { CSSTransitionGroup } from 'react-transition-group';
+import classnames from 'classnames';
+import styled from 'styled-components';
+import hexRgb from 'hex-rgb';
+import Title from '../Sidebar/Title';
```
Comment (GuillaumeAmat) [inline src/components/Modal/Modal.jsx]: I think we will have to make a choice. Create a `Modal.Title` component or a generic `Title` one used in both `Modal` and `Sidebar`.

I would prefer the generic choice, what's your opinion?

File: src/components/Modal/Modal.jsx [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+const StyledAside = styled.aside`
+  z-index: 1000;
+  top: 0;
+  bottom: 0;
+  left: 0;
+  right: 0;
+  padding: 10%;
+  background: ${props => {
+    const colorAsAnArray = hexRgb(props.theme.modal.backgroundColor);
+    colorAsAnArray.push(0.8);
```
Comment (GuillaumeAmat) [inline src/components/Modal/Modal.jsx]: `0.8` is a bit low for me. I have trouble to read the content whereas I have a good sight. What about `0.9`?

File: src/components/Modal/Modal.jsx [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+    return `rgba(${colorAsAnArray.join(',')})`;
+  }};
+  color: white;
```
Comment (GuillaumeAmat) [inline src/components/Modal/Modal.jsx]: That should be set in the theme.

File: src/helpers/themes.js [hunk 4]
```diff
@@ -91,6 +91,9 @@ export function buildDarkThemeConfig(baseConfig, colors, colorVariant) {
           activeBorderColor: color1
         }
       }
+    },
+    modal: {
```
Comment (GuillaumeAmat) [inline src/helpers/themes.js:102]: You forgot to add that `modal` section in the `buildLightThemeConfig` function below.

File: stories/Modal.jsx [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+  BlueTheme,
+  SkyTheme,
+  TurquoiseTheme,
+  GreenTheme,
+  Button,
+  Modal
+} from '../src/index';
+
+/*
+  *** WARNING ***
```
Comment (GuillaumeAmat) [inline stories/Modal.jsx:39]: Yep, we have to make a choice :)

**Review by GuillaumeAmat — CHANGES_REQUESTED** [review]

File: src/components/Modal/Modal.jsx (hunk 1 above)
Comment (bleucitron) [reply src/components/Modal/Modal.jsx]: I 100% agree with the generic approach.

It should be in a different PR though, I suggest this matter shouldn't block the merging of the current PR.

File: src/components/Modal/Modal.jsx (hunk 2 above)
Comment (bleucitron) [reply src/components/Modal/Modal.jsx]: Sure.

File: src/helpers/themes.js (hunk 4 above)
Comment (bleucitron) [reply src/helpers/themes.js:102]: 😑

File: src/components/Modal/Modal.jsx (hunk 3 above)
Comment (bleucitron) [reply src/components/Modal/Modal.jsx]: I agree

File: src/components/Modal/Modal.jsx (hunk 1 above)
Comment (GuillaumeAmat) [reply src/components/Modal/Modal.jsx]: I Agree

**Comment by bleucitron** [comment]
Please don't merge ! I still have modifs in the end.

File: src/themes/Default.jsx [hunk 6]
```diff
@@ -289,6 +289,9 @@ export const config = {
       backgroundColor: colors.red5,
       borderColor: colors.red5
     }
+  },
+  modal: {
+    backgroundColor: colors.black
```
Comment (GuillaumeAmat) [inline src/themes/Default.jsx]: It's missing the color item ;)

Also, the default theme modal colors should be the same as the root colors of that theme. So the modal item should be:

```
  modal: {
    color: colors.anthracite1,
    backgroundColor: colors.white,
  }

```

Comment (bleucitron) [reply src/themes/Default.jsx]: Given the fact that the `backgroundColor` of the modal is about the same color of the theme, but with opacity 90%, what you suggest makes the text unreadable.
I will leave `white` for now.

Comment (GuillaumeAmat) [reply src/themes/Default.jsx]: No it's not ^^

The modal itself with anthracite text on white background is clearly readable. And the theme of the underneath app can/should have another theme than the modla theme.

Eg: A red app opens a default (white) modal

Furthermore, if the developer wants to use a dark modal, he can choose to use the Anthracite theme, but the default have to stay white as the other components behaviors.

Comment (bleucitron) [reply src/themes/Default.jsx]: What about
```
    modal: {
      color: colors.black, // for default theme
      color: colors.black, // for light themes
      color: colors.white, // for dark themes
      backgroundColor: makeTransparent(color1, 0.9)
    }
```
?

Comment (GuillaumeAmat) [reply src/themes/Default.jsx]: I don't understand why you don't want to apply the inplace themes ^^ I'm certainly missing a point but I can't figure it out.

All the themes have a background color and a text color and it's working well everywhere. I tested it on the Modal component and it's the same result.

So for the default theme it should be:
```
modal: {
  color: colors.anthracite1,
  backgroundColor: colors.white,
}
```

Comment (GuillaumeAmat) [reply src/themes/Default.jsx]: For the dark one:
```
modal: {
  color: colors.white,
  backgroundColor: color2,
}
```

And for the light one:
```
modal: {
  color: dark27Color1,
  backgroundColor: color2,
}
```

Comment (bleucitron) [reply src/themes/Default.jsx]: Ok, actually I DO want to apply the inplace themes, but I totally missed out how your color defaults were set up for each theme, and tried to figure out a 'ok' styling.
Sorry about that.

Comment (GuillaumeAmat) [reply src/themes/Default.jsx]: Oh ok! ^^ No problem!

Comment (bleucitron) [reply src/themes/Default.jsx]: (keeping in mind that I apply 90% transparency, which will change slightly the theme color)

Comment (GuillaumeAmat) [reply src/themes/Default.jsx]: Yep

**Review by GuillaumeAmat — APPROVED** [review]

**Comment by GuillaumeAmat** [comment]
@iOiurson Oh crap, I think we forgot something again... A cross to close the modal ^^'

**Comment by bleucitron** [comment]
I didn't forget, I chose not to add one :).
In my opinion, a modal **requires** an action of the user.
And since we have some much space to add text and buttons, I would suggest that we leave entirely the closing of the `Modal` component to its parent (`Root`, or a Modal Manager) via a Button and the `onClose` prop.

**Comment by GuillaumeAmat** [comment]
It makes sense :)

Thank you for your patience ;)

**Pull request merged**